src/transpile/fusion.hpp

```diff
@@ -32,6 +32,645 @@ using oplist_t = std::vector<op_t>;
 using opset_t = Operations::OpSet;
 using reg_t = std::vector<uint_t>;
 
+class FusionMethod {
+public:
+  // Return name of method
+  virtual std::string name() = 0;
+
+  virtual bool support_diagonal() const = 0;
+
+  // Aggregate a subcircuit of operations into a single operation
+  virtual op_t generate_operation(std::vector<op_t>& fusioned_ops, bool diagonal = false) const {
```

Comment:
Does diagonal mean all ops in fusioned_ops are diagonal, or that the output is diagonal even if the ops are not?
If its the first we should have a specialized simulation method that doesn't require building the full square matrix.

Reply:
It is the second because CNOT-diagonal-CNOT patttern is detected.

I implemented [more efficient version](https://github.com/hhorii/qiskit-aer/blob/9f593798e0a9d9be4bcc7549797d7a3b0349a8fe/src/transpile/fusion_method.hpp#L160-L180) to generate a diagonal matrix but performance improvement was very limited for generation of small diagonal matrix.